pyomo/contrib/appsi/writers/nl_writer.py

```diff
@@ -223,3 +230,33 @@ def get_ordered_cons(self):
 
     def get_active_objective(self):
         return self._objective
+
+    def _set_pyomo_amplfunc_env(self):
```

Comment:
This seems to be block-copied (and mostly uncovered) from `ampl_.py`.  Can you extract that code into a public function there and directly reuse it here instead of copying it?

Reply:
Done.